In [1]:
import pgeocode

# Initialize the Nominatim object for Germany
nomi = pgeocode.Nominatim('DE')

# Query all postal codes for a specific city (e.g., "Berlin")
#city_data = nomi.query_location('Baden-Württemberg')  # Replace with a specific postal code
# print(city_data)
zip_data = nomi.query_postal_code('76669') 
#zip_data1 = nomi.query_location('Heidelberg',top_k=5) 
print(zip_data)
print(dir(nomi))

postal_code                   76669
country_code                     DE
place_name            Bad Schönborn
state_name        Baden-Württemberg
state_code                       BW
county_name        Karlsruhe Region
county_code                    82.0
community_name            Karlsruhe
community_code               8215.0
latitude                    49.2069
longitude                    8.6436
accuracy                        4.0
Name: 0, dtype: object
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_data', '_data_frame', '_data_path', '_fuzzy_search', '_get_data', '_index_postal_codes', '_normalize_postal_code', '_str_contains_search', 'country', 'query_location', 'query_postal_

In [13]:
admin_level_country = 2
admin_level_bundesland = 4
admin_level_regierungsbezirk = 5
admin_level_landkreis = 6
admin_level_stadt = 8

In [146]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Berlin"
# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
area["ISO3166-1"="DE"]->.searchArea;
relation["name"~"^({name})$",i][boundary=administrative](area.searchArea);
out center;
"""

location_annotation = {
    "city": ["Berlin", "Heidelberg", "Wiesloch", "Schwetzingen", "Bensheim"],  # admin level <= 8
}

# Run the query
result = api.query(query)

# Process relations
for relation in result.relations:
    print(f"Relation ID: {relation.id}")
    print(f"Latitude: {relation.center_lat}")
    print(f"Longitude: {relation.center_lon}")
    print("Tags:")
    for tag_key, tag_value in relation.tags.items():
        print(f"  {tag_key}: {tag_value}")
    print("\n")

Relation ID: 62422
Latitude: 52.5068768
Longitude: 13.4247530
Tags:
  ISO3166-2: DE-BE
  TMC:cid_58:tabcd_1:Class: Area
  TMC:cid_58:tabcd_1:LCLversion: 12.0
  TMC:cid_58:tabcd_1:LocationCode: 266
  admin_level: 4
  alt_name:vi: Béc-lin
  boundary: administrative
  coat_of_arms: File:Coat of arms of Berlin.svg
  contact:facebook: http://www.facebook.com/Berlin
  de:amtlicher_gemeindeschluessel: 11000000
  de:place: city
  de:regionalschluessel: 110000000000
  flag: File:Flag of Berlin.svg
  geographical_region: Barnim;Berliner Urstromtal;Teltow;Nauener Platte
  license_plate_code: B
  name: Berlin
  name:af: Berlyn
  name:als: Berlin
  name:am: በርሊን
  name:an: Berlín
  name:ang: Berlin
  name:ar: برلين
  name:arc: ܒܪܠܝܢ
  name:ast: Berlín
  name:az: Berlin
  name:ba: Берлин
  name:bar: Berlin
  name:bat-smg: Berlīns
  name:be: Берлін
  name:be-tarask: Бэрлін
  name:bg: Берлин
  name:bi: Berlin
  name:bn: বার্লিন
  name:bo: པེ་ལིང་
  name:br: Berlin
  name:bs: Berlin
  name:ca: Berlín
 

In [ ]:
import overpy
import re

# Initialize the Overpass API
api = overpy.Overpass()

# List of locations you want to query
locations = [
    "Berlin", "Heidelberg", "Wiesloch", "Schwetzingen", "Bensheim",
    "Karlsruhe", "Deutschland", "Österreich", "Italien",
    "Rhein-Neckar-Kreis", "Baden-Württemberg", "Odenwaldkreis"
]

# Escape location names to handle any special regex characters
escaped_locations = [re.escape(location) for location in locations]

# Create a regex pattern for the location names
locations_pattern = '|'.join(escaped_locations)

# Construct the Overpass QL query with proper OR conditions
query = f"""
[out:json][timeout:60];
(
  relation["name"~"^({locations_pattern})$",i]["boundary"="administrative"];
  relation["name:de"~"^({locations_pattern})$",i]["boundary"="administrative"];
);
out center tags;
"""

try:
    # Run the query
    result = api.query(query)

    # Intermediate dictionary to track the highest admin_level for each location
    # Key: location name, Value: admin_level (integer)
    location_admin_levels = {}

    # Process each relation (location)
    for relation in result.relations:
        # Prefer 'name:de' over 'name' if available
        location_name = relation.tags.get("name:de") or relation.tags.get("name", "Unknown")
        location_name = location_name.strip()

        admin_level_str = relation.tags.get("admin_level")

        # Convert admin_level to integer if possible, else assign 0
        try:
            admin_level = int(admin_level_str) if admin_level_str is not None else 0
        except ValueError:
            admin_level = 0

        # Update the highest admin_level for the location
        if location_name in location_admin_levels:
            if admin_level < location_admin_levels[location_name]:
                location_admin_levels[location_name] = admin_level
        else:
            location_admin_levels[location_name] = admin_level

    # Initialize the final dictionary with admin_level as keys and list of locations as values
    admin_levels_dict = {}

    for location in locations:
        # Retrieve the admin_level for the location, default to 0 if not found
        admin_level = location_admin_levels.get(location, 0)

        # Initialize the list for this admin_level if not already present
        if admin_level not in admin_levels_dict:
            admin_levels_dict[admin_level] = []

        # Add the location to the corresponding admin_level
        admin_levels_dict[admin_level].append(location)

    # Print the dictionary in a structured format
    for admin_level in sorted(admin_levels_dict.keys()):
        print(f"Admin Level: {admin_level}")
        print(f"  Locations: {', '.join(admin_levels_dict[admin_level])}\n")

except overpy.exception.OverpassTooManyRequests:
    print("Error: Too many requests. Please try again later.")
except overpy.exception.OverpassGatewayTimeout:
    print("Error: The Overpass API timed out. Please try again later.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")


Admin Level: 2
  Locations: Deutschland, Österreich, Italien

Admin Level: 4
  Locations: Berlin, Baden-Württemberg

Admin Level: 6
  Locations: Heidelberg, Karlsruhe, Rhein-Neckar-Kreis, Odenwaldkreis

Admin Level: 8
  Locations: Wiesloch, Schwetzingen, Bensheim



In [147]:
print(admin_levels_dict)

{4: ['Berlin', 'Baden-Württemberg'], 6: ['Heidelberg', 'Karlsruhe', 'Rhein-Neckar-Kreis', 'Odenwaldkreis'], 8: ['Wiesloch', 'Schwetzingen', 'Bensheim'], 2: ['Deutschland', 'Österreich', 'Italien']}


In [119]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Heidelberg"
admin_level = 2
# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
area["ISO3166-1"="DE"]->.searchArea;
relation["note"~"{name}"][boundary=postal_code](area.searchArea);
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for relation in result.relations:
    print(f"Relation ID: {relation.id}")
    print(f"Latitude: {relation.center_lat}")
    print(f"Longitude: {relation.center_lon}")
    print("Tags:")
    for tag_key, tag_value in relation.tags.items():
        print(f"  {tag_key}: {tag_value}")
    print("\n")


Relation ID: 1186414
Latitude: 49.3938369
Longitude: 8.7306991
Tags:
  boundary: postal_code
  note: 69117 Heidelberg
  postal_code: 69117
  postal_code_level: 8
  ref: 69117
  type: boundary


Relation ID: 1186421
Latitude: 49.4179176
Longitude: 8.6194847
Tags:
  boundary: postal_code
  note: 69123 Heidelberg
  postal_code: 69123
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010
  type: boundary


Relation ID: 1186427
Latitude: 49.4230453
Longitude: 8.7558743
Tags:
  boundary: postal_code
  note: 69118 Heidelberg
  postal_code: 69118
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010
  type: boundary


Relation ID: 1186432
Latitude: 49.4197653
Longitude: 8.6908328
Tags:
  boundary: postal_code
  note: 69120 Heidelberg
  postal_code: 69120
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010


In [47]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Define the latitude, longitude, radius (in meters), and admin_level
latitude = 49.4058478 
longitude = 8.6835826 
radius = 10_000 
admin_level = 6  

# Overpass QL query to search around a point for relations within a radius
query = f"""
[out:json][timeout:60];
(
  relation[boundary=postal_code](around:{radius},{latitude},{longitude});
);
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for relation in result.relations:
    print(f"Relation ID: {relation.id}")
    print(f"Latitude: {relation.center_lat}")
    print(f"Longitude: {relation.center_lon}")
    print("Tags:")
    for tag_key, tag_value in relation.tags.items():
        print(f"  {tag_key}: {tag_value}")
    print("\n")


Relation ID: 1183836
Latitude: 49.4350556
Longitude: 8.5355200
Tags:
  boundary: postal_code
  note: 68219 Mannheim
  postal_code: 68219
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010
  type: boundary


Relation ID: 1183839
Latitude: 49.3776792
Longitude: 8.5829441
Tags:
  boundary: postal_code
  note: 68723 Schwetzingen, Oftersheim, Plankstadt
  postal_code: 68723
  postal_code_level: 8
  type: boundary


Relation ID: 1183842
Latitude: 49.3661581
Longitude: 8.5261267
Tags:
  boundary: postal_code
  note: 68775 Ketsch
  postal_code: 68775
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010
  type: boundary


Relation ID: 1183844
Latitude: 49.4510438
Longitude: 8.6009605
Tags:
  boundary: postal_code
  note: 68535 Edingen-Neckarhausen
  postal_code: 68535
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_

In [69]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Define the latitude, longitude, radius (in meters), and admin_level
latitude = 49.4058478
longitude = 8.6835826
radius = 100

# Overpass QL query to search around a point for ways with the "highway" tag within a radius
query = f"""
[out:json][timeout:60];
way["highway"]["name"](around:{radius},{latitude},{longitude});
out body;
"""

# Run the query
result = api.query(query)

# Process ways
for way in result.ways:
    # print(f"Way ID: {way.id}")
    # print("Tags:")
    # for tag_key, tag_value in way.tags.items():
    #     print(f"  {tag_key}: {tag_value}")
    print(f"  Name: {way.tags['name']}")


  Name: Römerkreis
  Name: Ringstraße
  Name: Gaswerkstraße
  Name: Römerstraße
  Name: Römerkreis
  Name: Römerkreis
  Name: Römerkreis
  Name: Ringstraße
  Name: Kurfürsten-Anlage
  Name: Römerkreis
  Name: Römerstraße
  Name: Kurfürsten-Anlage
  Name: Ringstraße
  Name: Kurfürsten-Anlage
  Name: Römerstraße
  Name: Kurfürsten-Anlage
  Name: Kurfürsten-Anlage
  Name: Römerkreis
  Name: Römerstraße
  Name: Römerkreis
  Name: Alte Eppelheimer Straße
  Name: Römerkreis
  Name: Kurfürsten-Anlage


# Get Neighboring Countries

In [ ]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

name = "Deutschland"
admin_level = 2

# Overpass QL query to get neighboring countries of Germany
query = f"""
[out:json][timeout:60];
// Get the relation for the Location Area
relation["name"="{name}"]["boundary"="administrative"]["admin_level"="{admin_level}"]->.country;

// Get all ways that are members of this relation
way(r.country)->.country_ways;

// Find all relations that share these boundary ways
relation(bw.country_ways)["boundary"="administrative"]["type"="boundary"]["name:de"]["admin_level"="{admin_level}"]->.neighbors;

// Output the neighboring relations
relation.neighbors;
out center;
"""

# Run the query
result = api.query(query)

# Process relations
print(f"Neighboring Countries to {name}:")
for relation in result.relations:
    # print(f"Relation ID: {relation.id}")
    # print(f"Latitude: {relation.center_lat}")
    # print(f"Longitude: {relation.center_lon}")
    print(relation.tags.get('name:de'))
    # for tag_key, tag_value in relation.tags.items():
    #     print(f"  {tag_key}: {tag_value}")




Neighboring Countries to Germany:
Österreich
Polen
Dänemark
Deutschland
Tschechien
Schweiz
Belgien
Luxemburg
Frankreich
Niederlande


# Get Neighboring States

In [103]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Berlin"
admin_level = 4
# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
//Set search area to Germany
area["ISO3166-1"="DE"]->.searchArea;
// Get the relation for the Location Area
relation["name"~"{name}"]["boundary"="administrative"]["admin_level"="{admin_level}"](area.searchArea)->.bundesland;

// Get all ways that are members of this relation
way(r.bundesland)->.bundesland_ways;

// Find all relations that share these boundary ways
relation(bw.bundesland_ways)["boundary"="administrative"]["type"="boundary"]["name:de"]["admin_level"="{admin_level}"](area.searchArea)->.neighbors;

// Output the neighboring relations
relation.neighbors;
out center;
"""

# Run the query
result = api.query(query)

# Process relations
print(f"Neighboring Countries to {name}:")
for relation in result.relations:
    # print(f"Relation ID: {relation.id}")
    # print(f"Latitude: {relation.center_lat}")
    # print(f"Longitude: {relation.center_lon}")
    print(relation.tags.get('name:de'))
    # for tag_key, tag_value in relation.tags.items():
    #     print(f"  {tag_key}: {tag_value}")


Neighboring Countries to Berlin:
Berlin
Brandenburg


# Get Landkreise inside of State

In [121]:
import overpy
admin_level_country = 2
admin_level_bundesland = 4
admin_level_regierungsbezirk = 5
admin_level_landkreis = 6
admin_level_stadt = 8
# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Baden-Württemberg"

# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
area["ISO3166-1"="DE"]->.searchArea;
relation["name"~"{name}"]["boundary"="administrative"]["admin_level"="{admin_level_bundesland}"](area.searchArea)->.bundesland;

// convert relations object into an area object
.bundesland map_to_area -> .bundeslandArea;

// Find all relations that are a member of this area
relation["boundary"="administrative"]["admin_level"="{admin_level_landkreis}"](area.bundeslandArea)->.landkreise;

// Output the neighboring relations
relation.landkreise;
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for relation in result.relations:
    # print(f"Relation ID: {relation.id}")
    # print(f"Latitude: {relation.center_lat}")
    # print(f"Longitude: {relation.center_lon}")
    # print("Tags:")
    # for tag_key, tag_value in relation.tags.items():
    #     print(f"  {tag_key}: {tag_value}")
    # print("\n")
    print(f"  Name: {relation.tags['name']}")

  Name: Baden-Baden
  Name: Landkreis Rottweil
  Name: Landkreis Freudenstadt
  Name: Landkreis Rastatt
  Name: Landkreis Karlsruhe
  Name: Hohenlohekreis
  Name: Rems-Murr-Kreis
  Name: Landkreis Würzburg
  Name: Pforzheim
  Name: Ulm
  Name: Karlsruhe
  Name: Landkreis Ludwigsburg
  Name: Main-Tauber-Kreis
  Name: Rhein-Neckar-Kreis
  Name: Landkreis Schwäbisch Hall
  Name: Landkreis Lörrach
  Name: Landkreis Calw
  Name: Landkreis Konstanz
  Name: Ortenaukreis
  Name: Neckar-Odenwald-Kreis
  Name: Landkreis Tuttlingen
  Name: Schwarzwald-Baar-Kreis
  Name: Mannheim
  Name: Ostalbkreis
  Name: Landkreis Waldshut
  Name: Landkreis Böblingen
  Name: Enzkreis
  Name: Landkreis Heilbronn
  Name: Heilbronn
  Name: Freiburg im Breisgau
  Name: Heidelberg
  Name: Landkreis Emmendingen
  Name: Landkreis Breisgau-Hochschwarzwald
  Name: Stuttgart
  Name: Landkreis Reutlingen
  Name: Landkreis Tübingen
  Name: Zollernalbkreis
  Name: Alb-Donau-Kreis
  Name: Landkreis Biberach
  Name: Landkreis

# Get City inside of Landkreis

In [4]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Rhein-Neckar-Kreis"

# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
area["ISO3166-1"="DE"]->.searchArea;
relation["name"~"{name}"]["boundary"="administrative"]["admin_level"="{admin_level_landkreis}"](area.searchArea)->.landkreis;

// convert relations object into an area object
.landkreis map_to_area -> .landkreisArea;

// Find all relations that are a member of this area
relation["boundary"="administrative"]["admin_level"="{admin_level_stadt}"](area.landkreisArea)->.stadt;

// Output the neighboring relations
relation.stadt;
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for relation in result.relations:
    print(f"Relation ID: {relation.id}")
    # print(f"Latitude: {relation.center_lat}")
    # print(f"Longitude: {relation.center_lon}")
    # print("Tags:")
    # for tag_key, tag_value in relation.tags.items():
    #     print(f"  {tag_key}: {tag_value}")
    # print("\n")
    print(f"  Name: {relation.tags['name']}")

NameError: name 'admin_level_landkreis' is not defined

# Get Sreet inside of City

In [ ]:
import overpy
admin_level_stadt = 8
# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Mühlhausen"

# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
area["ISO3166-1"="DE"]->.searchArea;
relation["name"="{name}"]["boundary"="administrative"]["admin_level"="{admin_level_stadt}"](area.searchArea)->.stadt; // relation(453374)->.stadt;

// convert relations object into an area object
.stadt map_to_area -> .stadtArea;

// Find all relations that are a member of this area
way["highway"]["name"](area.stadtArea)->.strasse;

// Output the neighboring relations
way.strasse;
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for ways in result.ways:
    # print(f"Relation ID: {ways.id}")
    # print(f"Latitude: {ways.center_lat}")
    # print(f"Longitude: {ways.center_lon}")
    # print("Tags:")
    # for tag_key, tag_value in ways.tags.items():
    #     print(f"  {tag_key}: {tag_value}")
    # print("\n")
    print(f"  Name: {ways.tags['name']}")

# Is relation contained in another relation? (clustering operation)

In [49]:
admin_level_country = 2
admin_level_bundesland = 4
admin_level_regierungsbezirk = 5
admin_level_landkreis = 6
admin_level_verwaltungsgemeinschaft = 7
admin_level_stadt = 8
admin_level_stadtteil = 9

import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Names of the district and city
district_name = "Neuenheim"
city_name = "Heidelberg"

# Overpass QL query to check if the district is within the city
query = f"""
[out:json][timeout:60];

// Get the area ID of the city
area[name="{city_name}"]["boundary"="administrative"]->.cityArea;

// Now get the relation of the district within that area
relation["name"="{district_name}"]["boundary"="administrative"](area.cityArea);

// Outputting only ids is an efficient way to determine if a match exists
out ids;
"""

# Run the query
result = api.query(query)

# Check if the district is within the city
if result.relations:
    print(f"The district '{district_name}' is within the city '{city_name}'.")
else:
    print(f"The district '{district_name}' is not within the city '{city_name}'.")


The district 'Neuenheim' is within the city 'Heidelberg'.


# Determine Bundesland of location

In [42]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# The place name you want to search for
place_name = "Bensheim"

# First, fetch the place and get its center point
query_place = f"""
[out:json][timeout:25];
area["ISO3166-1"="DE"]->.searchArea;
// Fetch the relation for the place and get its center point
relation["name"="{place_name}"]["boundary"="administrative"](area.searchArea);
out center;
"""

# Run the query to get the place's center point
result_place = api.query(query_place)

if result_place.relations:
    place_relation = result_place.relations[0]
    center_lat = place_relation.center_lat
    center_lon = place_relation.center_lon

    # Now, use the center point to find the Bundesland containing it
    query_bundesland = f"""
    [out:json][timeout:25];

    // Use the center point to find containing administrative boundaries
    is_in({center_lat},{center_lon});
    // Select relations that are administrative boundaries at admin_level=4 (Bundesländer)
    relation(pivot)["boundary"="administrative"]["admin_level"="4"];
    out tags;
    """

    # Run the query to find the Bundesland
    result_bundesland = api.query(query_bundesland)

    if result_bundesland.relations:
        # Assuming there's only one Bundesland containing the point
        bundesland_relation = result_bundesland.relations[0]
        state_name = bundesland_relation.tags.get("name")
        print(f"The place '{place_name}' is in the Bundesland: {state_name}")
    else:
        print(f"The Bundesland for '{place_name}' could not be found.")
else:
    print(f"The place '{place_name}' could not be found.")




The place 'Bensheim' is in the Bundesland: Hessen


# Cluster locations into Hierarchy

In [ ]:
location_annotation = {
    "country": ["Deutschland", "Österreich", "Italien"],                                                # admin level = 2 / Land
    "state": ["Rhein-Neckar-Kreis", "Baden-Württemberg", "Odenwaldkreis"],                              # admin level 3 < x < 7 / 4:Bundesland, 5:Regierungsbezirk, 6:Landkreis
    "city": ["Berlin", "Heidelberg", "Wiesloch", "Schwetzingen", "Bensheim"],                           # admin level = <= 8 / 8:Stadt, 9:Stadtteil, 10:Ortsteil 
    "zip": ["10178", "69115", "69168", "68723", "64625"],                                               # admin level = <= 8 / 8:Stadt, 9:Stadtteil, 10:Ortsteil 
    "street": ["Spandauer Str. 3", "Bergstraße 3", "Carl-Theodor-Straße 13", "Fabrikstraße 10"],
}

['Deutschland', 'Österreich', 'Italien']


In [ ]:
location_annotation = {
    "country": ["Deutschland", "Österreich", "Italien"],                                                # admin level = 2 / Land
    "state": ["Rhein-Neckar-Kreis", "Baden-Württemberg", "Odenwaldkreis"],                              # admin level 3 < x < 7 / 4:Bundesland, 5:Regierungsbezirk, 6:Landkreis
    "city": ["Berlin", "Heidelberg", "Wiesloch", "Schwetzingen", "Bensheim"],                           # admin level = <= 8 / 8:Stadt, 9:Stadtteil, 10:Ortsteil 
    "street": ["Spandauer Str. 3", "Bergstraße 3", "Carl-Theodor-Straße 13", "Fabrikstraße 10"],
}

In [82]:
admin_level_country = 2
admin_level_bundesland = 4
admin_level_regierungsbezirk = 5
admin_level_landkreis = 6
admin_level_stadt = 8



def is_child(parent_name, parent_admin_level, child_name, child_admin_level, api):
    """
    This function uses the Overpass API to check whether a given child region is contained
    within the boundaries of a specified parent region. 

    Parameters
    ----------
    parent_name : str
        The name of the parent administrative region.
    parent_admin_level : int
        The administrative level of the parent region (e.g., 4 for states in Germany).
    child_name : str
        The name of the child administrative region.
    child_admin_level : Union[int, List[int]]
        Either a single administrative level (int) or a range as a list of two integers [lower, upper].
    api : overpy.Overpass
        An instance of the Overpass API for querying OpenStreetMap data.

    Returns
    -------
    bool
        True if the child administrative region is within the parent region; False otherwise.
    """
    try:
        # Handle single value or range for child_admin_level
        if isinstance(child_admin_level, int):
            admin_level_filter = f'["admin_level"="{child_admin_level}"]'
        elif isinstance(child_admin_level, list) and len(child_admin_level) == 2:
            admin_level_filter = f'["admin_level"~"^[{child_admin_level[0]}-{child_admin_level[1]}]$"]'
        else:
            raise ValueError("child_admin_level must be an int or a list of two integers.")

        # Overpass QL query
        query = f"""
        [out:json];
        area["ISO3166-1"="DE"]->.searchArea;
        relation["name"~"{parent_name}"]["boundary"="administrative"]["admin_level"="{parent_admin_level}"](area.searchArea)->.parent;

        // convert relations object into an area object
        .parent map_to_area -> .parentArea;

        // Find all relations that are a member of this area
        relation["name"~"{child_name}"]["boundary"="administrative"]{admin_level_filter}(area.parentArea)->.child;

        // Output the neighboring relations
        relation.child;
        out ids;
        """

        # Run the query
        result = api.query(query)
        
        # Return True if any child relations were found
        return bool(result.relations)

    except overpy.exception.OverpassBadRequest as e:
        logging.error(f"Bad Overpass Query: {e}")
        return False
    except ValueError as e:
        logging.error(f"Invalid Input: {e}")
        return False
    except Exception as e:
        logging.error(f"Unexpected Error: {e}")
        return False

    
import logging   
import overpy
# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
parent_name = "Baden-Württemberg"
parent_admin_level = 4
child_name = "Heidelberg"
child_admin_level_range = [6,8] # list or int
  
print(is_child(parent_name, parent_admin_level, child_name, child_admin_level_range, api))

True


# Get City name for zip

In [26]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Specify the postal code for the query
postal_code = "64625"  # Replace with the desired postal code

# Overpass QL query to retrieve coordinates for the given postal code
query = f"""
[out:json][timeout:60];
area["boundary"="postal_code"]["postal_code"="{postal_code}"];
out center;
"""


result = api.query(query)

# Process areas
for area in result.areas:
    print(f"Note: {area.tags['note']}")
    print(area.tags)

Note: 64625 Bensheim
{'boundary': 'postal_code', 'note': '64625 Bensheim', 'postal_code': '64625', 'postal_code_level': '8', 'type': 'boundary'}


# Get surrogate for zip 

In [22]:
import overpy

# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
name = "Heidelberg"
admin_level = 2
# Overpass QL query to get basic information about a city
query = f"""
[out:json][timeout:60];
area["ISO3166-1"="DE"]->.searchArea;
relation["note"~"{name}"][boundary=postal_code](area.searchArea);
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for relation in result.relations:
    print(f"Relation ID: {relation.id}")
    print(f"Latitude: {relation.center_lat}")
    print(f"Longitude: {relation.center_lon}")
    print("Tags:")
    for tag_key, tag_value in relation.tags.items():
        print(f"  {tag_key}: {tag_value}")
    print("\n")

Relation ID: 1186414
Latitude: 49.3938369
Longitude: 8.7306991
Tags:
  boundary: postal_code
  note: 69117 Heidelberg
  postal_code: 69117
  postal_code_level: 8
  ref: 69117
  type: boundary


Relation ID: 1186421
Latitude: 49.4179176
Longitude: 8.6194847
Tags:
  boundary: postal_code
  note: 69123 Heidelberg
  postal_code: 69123
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010
  type: boundary


Relation ID: 1186427
Latitude: 49.4230453
Longitude: 8.7558743
Tags:
  boundary: postal_code
  note: 69118 Heidelberg
  postal_code: 69118
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010
  type: boundary


Relation ID: 1186432
Latitude: 49.4197653
Longitude: 8.6908328
Tags:
  boundary: postal_code
  note: 69120 Heidelberg
  postal_code: 69120
  postal_code_level: 8
  source: http://wiki.openstreetmap.org/wiki/Import/Catalogue/Postleitzahlen_Deutschland_2010


In [ ]:
admin_level_country = 2
admin_level_bundesland = 4
admin_level_regierungsbezirk = 5
admin_level_landkreis = 6
admin_level_stadt = 8

location_annotation = {
    "country": ["Deutschland", "Österreich", "Italien"],  # admin level = 2 / Land
    "state": ["Rhein-Neckar-Kreis", "Baden-Württemberg", "Odenwaldkreis"],  # admin level 4
    "city": ["Berlin", "Heidelberg", "Wiesloch", "Schwetzingen", "Bensheim"],  # admin level <= 8
}

def is_child(parent_id, child_name, child_admin_level, api):
    """
    This function uses the Overpass API to check whether a given child region is contained
    within the boundaries of a specified parent region. 

    Parameters
    ----------
    parent_name : str
        The name of the parent administrative region.
    parent_admin_level : int
        The administrative level of the parent region (e.g., 4 for states in Germany).
    child_name : str
        The name of the child administrative region.
    child_admin_level : Union[int, List[int]]
        Either a single administrative level (int) or a range as a list of two integers [lower, upper].
    api : overpy.Overpass
        An instance of the Overpass API for querying OpenStreetMap data.

    Returns
    -------
    bool
        True if the child administrative region is within the parent region; False otherwise.
    """
    try:
        # Handle single value or range for child_admin_level
        if isinstance(child_admin_level, int):
            admin_level_filter = f'["admin_level"="{child_admin_level}"]'
        elif isinstance(child_admin_level, list) and len(child_admin_level) == 2:
            admin_level_filter = f'["admin_level"~"^[{child_admin_level[0]}-{child_admin_level[1]}]$"]'
        else:
            raise ValueError("child_admin_level must be an int or a list of two integers.")

        # Overpass QL query
        query = f"""
        [out:json];
        //Use the ID of the parent to specify Search Area
        relation({parent_id})->.parent;

        // convert relations object into an area object
        .parent map_to_area -> .parentArea;

        // Find all relations that are a member of this area
        relation["name"~"{child_name}"]["boundary"="administrative"]{admin_level_filter}(area.parentArea)->.child;

        // Output the neighboring relations
        relation.child;
        out ids;
        """

        # Run the query
        result = api.query(query)
        
        # Return True if any child relations were found
        return result.relations[0].id if result.relations else None

    except overpy.exception.OverpassBadRequest as e:
        logging.error(f"Bad Overpass Query: {e}")
        return False
    except ValueError as e:
        logging.error(f"Invalid Input: {e}")
        return False
    except Exception as e:
        logging.error(f"Unexpected Error: {e}")
        return False

    
import logging   
import overpy
# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
parent_id = 285864
child_name = "Heidelberg"
child_admin_level_range = [6,8] # list or int

print(is_child(parent_id, child_name, child_admin_level_range, api))

285864


# Determine Proper Location Admin Levels

In [ ]:
import overpy
import re
import logging
from anytree import Node, RenderTree

api = overpy.Overpass()
admin_level_city = 8

locations = [
    "Berlin", "Heidelberg", "Wiesloch", "Schwetzingen", "Bensheim",
    "Karlsruhe", "Rhein-Neckar-Kreis", "Baden-Württemberg", "Odenwaldkreis", "Mühlhausen"
]

street_locations = [
    "Spandauer Straße", "Bergstraße", "Carl-Theodor-Straße", "Fabrikstraße"
]

postal_codes = [69120, 10178, 69168, 68723, 64625]

def fetch_location_info(location_list):
    """
    Query Overpass for the given locations and return their best admin_level match.

    Parameters
    ----------
    location_list : list of str
        Location names to query.

    Returns
    -------
    dict
        {location_name: {'admin_level': int, 'id': int or None}}
    """
    escaped = [re.escape(loc) for loc in location_list]
    pattern = '|'.join(escaped)

    query = f"""
    [out:json][timeout:60];
    (
      relation["name"~"^({pattern})$",i]["boundary"="administrative"];
      relation["name:de"~"^({pattern})$",i]["boundary"="administrative"];
    );
    out center tags;
    """
    result = api.query(query)

    info = {}
    for r in result.relations:
        name = (r.tags.get("name:de") or r.tags.get("name", "Unknown")).strip()
        lvl_str = r.tags.get("admin_level", "")
        level = int(lvl_str) if lvl_str.isdigit() else 0
        if name not in info or level < info[name]['admin_level']:
            info[name] = {'admin_level': level, 'id': r.id}
    return info

def group_locations_by_admin_level(location_list, locations_data):
    """
    Group locations by their admin_level.

    Parameters
    ----------
    location_list : list of str
        The queried locations.
    locations_data : dict
        {location_name: {'admin_level': int, 'id': int or None}}

    Returns
    -------
    dict
        {admin_level: [{'name': str, 'id': int or None}, ...]}
    """
    grouped = {}
    for loc in location_list:
        data = locations_data.get(loc, {'admin_level': 0, 'id': None})
        level = data['admin_level']
        grouped.setdefault(level, []).append({'name': loc, 'id': data['id']})
    return grouped

def get_state(osm_id):
    """
    Retrieves the Bundesland (state) information for a given OSM relation ID.

    Parameters:
        osm_id (int): The OSM relation ID of the place.

    Returns:
        tuple: (state_name, state_osm_id) if found, else (None, None).
    """
    api = overpy.Overpass()

    # Query to fetch the relation using its OSM ID and get its center point
    query_place = f"""
    [out:json][timeout:25];
    relation({osm_id});
    out center;
    """

    try:
        result_place = api.query(query_place)
    except Exception as e:
        logging.error(f"Error fetching place with OSM ID {osm_id}: {e}")
        return None, None

    if not result_place.relations:
        logging.warning(f"No relation found for OSM ID {osm_id}.")
        return None, None

    place_relation = result_place.relations[0]

    # Ensure that the relation has a center point
    if not hasattr(place_relation, 'center_lat') or not hasattr(place_relation, 'center_lon'):
        logging.warning(f"Relation {osm_id} does not have a center point.")
        return None, None

    center_lat = place_relation.center_lat
    center_lon = place_relation.center_lon

    # Query to find the Bundesland containing the center point
    query_bundesland = f"""
    [out:json][timeout:25];
    is_in({center_lat},{center_lon});
    relation(pivot)["boundary"="administrative"]["admin_level"="4"];
    out ids tags;
    """

    try:
        result_bundesland = api.query(query_bundesland)
    except Exception as e:
        logging.error(f"Error fetching Bundesland for OSM ID {osm_id}: {e}")
        return None, None

    if not result_bundesland.relations:
        logging.warning(f"No Bundesland found containing OSM ID {osm_id}.")
        return None, None

    bundesland_relation = result_bundesland.relations[0]
    state_name = bundesland_relation.tags.get("name:de")
    state_osm_id = bundesland_relation.id

    if not state_name:
        logging.warning(f"Bundesland relation {state_osm_id} lacks a name tag.")
        return None, None

    return state_name, state_osm_id

def is_child(parent_id, child_name, api):
    """
    Check if child_name is within parent_id's area.

    Parameters
    ----------
    parent_id : int
        OSM relation ID of the parent area.
    child_name : str
        Name of the child area.
    api : overpy.Overpass
        Overpass API instance.

    Returns
    -------
    int or None
        Child's relation ID if found, else None.
    """
    query = f"""
    [out:json];
    relation({parent_id})->.parent;
    .parent map_to_area -> .parentArea;
    relation["name"~"^{re.escape(child_name)}$",i]["boundary"="administrative"](area.parentArea);
    out ids;
    """
    try:
        result = api.query(query)
        return result.relations[0].id if result.relations else None
    except (overpy.exception.OverpassBadRequest, Exception):
        return None

def build_hierarchy(locations_by_level):
    """
    Build a hierarchy of administrative nodes.
    After building the graph, iterate through the leaves and attempt to find a street for each.
    Once a single leaf with a matching street is found, add that street as a child to that leaf and stop.
    
    Parameters
    ----------
    locations_by_level : dict
        {admin_level: [{'name': str, 'id': int or None}, ...]}

    Returns
    -------
    dict
        Hierarchy nodes with the built hierarchy.
    """
    
    hierarchy_nodes = {}

    def is_child(parent_id, child_name,api):
        query = f"""
        [out:json];
        relation({parent_id})->.parent;
        .parent map_to_area -> .parentArea;
        relation["name"~"^{re.escape(child_name)}$",i]["boundary"="administrative"](area.parentArea);
        out ids;
        """
        try:
            result = api.query(query)
            return result.relations[0].id if result.relations else None
        except:
            return None

    def create_or_get_node(name, node_id, level):
        if node_id in hierarchy_nodes:
            return hierarchy_nodes[node_id]

        node = Node(name, id=node_id)
        node.admin_level = level
        hierarchy_nodes[node_id] = node

        if level == 2:
            return node

        higher_levels = sorted([l for l in locations_by_level if l < level], reverse=True)
        parent_found = False

        for hl in higher_levels:
            for p_info in locations_by_level[hl]:
                p_id = p_info['id']
                if p_id and is_child(p_id, name, api):
                    p_node = create_or_get_node(p_info['name'], p_id, hl)
                    node.parent = p_node
                    parent_found = True
                    break
            if parent_found:
                break

        # If no parent with admin_level=4 is found, fetch and create Bundesland
        if not parent_found and level > 4:
            state_name, state_id = get_state(node_id)
            if state_name and state_id:
                # Create Bundesland node if not already created
                if state_id not in hierarchy_nodes:
                    state_node = Node(state_name, id=state_id, admin_level=4)
                    hierarchy_nodes[state_id] = state_node
                else:
                    state_node = hierarchy_nodes[state_id]
                
                # Set as parent
                node.parent = state_node

        return node

    # Create all nodes
    for level in sorted(locations_by_level.keys(), reverse=True):
        for entry in locations_by_level[level]:
            create_or_get_node(entry['name'], entry['id'], level)

    def find_one_street_in_area(parent_id, street_name):
        query = f"""
        [out:json];
        relation({parent_id})->.parent;
        .parent map_to_area -> .parentArea;
        way["highway"]["name"~"^{re.escape(street_name)}$"](area.parentArea);
        out ids;
        """
        try:
            res = api.query(query)
            if res.ways:
                return (street_name, res.ways[0].id)
        except:
            pass
        return None

    # Identify all leaves
    leaves = [node for node in hierarchy_nodes.values() if node.is_leaf and node.id]

    # Try to find a street for each leaf until one is matched
    for leaf in leaves:
        for street in street_locations:
            street_info = find_one_street_in_area(leaf.id, street)
            if street_info:
                s_name, s_id = street_info
                Node(s_name, id=s_id, admin_level=99, parent=leaf) 
                # Once a single leaf with a street is found, we stop.
                return hierarchy_nodes

    return hierarchy_nodes

def update_tree_with_zip_codes(postal_codes, hierarchy_nodes, api):
    """
    Assign postal code attributes to nodes whose areas match the postal codes in the provided set.

    Parameters
    ----------
    postal_codes : set
        A set of postal codes to query using the Overpass API. These represent the areas of interest.
    hierarchy_nodes : dict
        A dictionary of node objects, where each node has the following attributes:
            - 'name' (str): The name of the node, used to match against area names from the query.
            - 'zip' (str): The postal code to be assigned to the node (this attribute will be updated).
            - 'is_root' (bool): Indicates if the node is a root node in the hierarchy.

    Returns
    -------
    dict
        The updated hierarchy_nodes dictionary with postal codes assigned to the 'zip' attribute of matching nodes.
    """
    # Create a regex pattern to match postal codes
    postal_code_filter = '|'.join(map(str, postal_codes))
    
    # Build the Overpass QL query to fetch areas with specified postal codes
    query = f"""
    area["boundary"="postal_code"]["postal_code"~"^({postal_code_filter})$"];
    out center;
    """
    
    try:
        # Execute the Overpass API query
        result = api.query(query)
    except:
        pass
    
    # Dictionary to store postal code names
    postal_code_names = {}
    
    # Use a regular expression to separate the numerical and textual parts
    pattern = re.compile(r'(\d+)\s*(.*)')  # d=digit s=white space .=any character 
    
    # Process each area returned by the Overpass API
    for area in result.areas:
        # Extract the 'note' tag which contains postal code information
        note = area.tags.get('note', '')
        
        # Use regex to separate the numerical and textual parts of the note
        match = pattern.match(note)
        if match:
            postal_code, name = match.groups()
            postal_code_names[postal_code] = name
    
    # Assign zip codes to nodes based on matching names
    for postal_code, name in postal_code_names.items():
        for node in hierarchy_nodes.values():
            # Check if the node's name contains the postal area name
            if node.name in name:
                node.zip = postal_code
    
    return hierarchy_nodes


try:
    locations_data = fetch_location_info(locations)
    locations_by_level = group_locations_by_admin_level(locations, locations_data)
    hierarchy_nodes = build_hierarchy(locations_by_level)
    hierarchy_nodes = update_tree_with_zip_codes(postal_codes, hierarchy_nodes, api)
    
    # Print hierarchy
    old_roots = [n for n in hierarchy_nodes.values() if n.is_root]
    for root in old_roots:
        for pre, _, node in RenderTree(root):
            # Print formatted tree with admin level and ID for better readability
            print(f"{pre}{node.name} (admin_level={node.admin_level}, id={node.id})")

except overpy.exception.OverpassTooManyRequests:
    print("Error: Too many requests. Please try again later.")
except overpy.exception.OverpassGatewayTimeout:
    print("Error: The Overpass API timed out. Please try again later.")
except Exception as e:
    print(f"Unexpected error: {e}")

Baden-Württemberg (admin_level=4, id=62611)
├── Rhein-Neckar-Kreis (admin_level=6, id=62577)
│   ├── Wiesloch (admin_level=8, id=453399)
│   │   └── Bergstraße (admin_level=99, id=5071422)
│   ├── Schwetzingen (admin_level=8, id=454165)
│   └── Mühlhausen (admin_level=8, id=437377)
├── Heidelberg (admin_level=6, id=285864)
└── Karlsruhe (admin_level=6, id=62518)
Hessen (admin_level=4, id=62650)
├── Bensheim (admin_level=8, id=537055)
└── Odenwaldkreis (admin_level=6, id=62596)
Berlin (admin_level=4, id=62422)


In [63]:
import copy

test_nodes = copy.deepcopy(hierarchy_nodes)
old_roots = [n for n in test_nodes.values() if n.is_root]
for root in old_roots:
    for pre, _, node in RenderTree(root):
        # Print formatted tree with admin level and ID for better readability
        zip_info = f", zip={node.zip}" if hasattr(node, 'zip') else ""
        print(f"{pre}{node.name} (admin_level={node.admin_level}, id={node.id}{zip_info})")

Baden-Württemberg (admin_level=4, id=62611)
├── Rhein-Neckar-Kreis (admin_level=6, id=62577)
│   ├── Wiesloch (admin_level=8, id=453399, zip=69168)
│   │   └── Bergstraße (admin_level=99, id=5071422)
│   ├── Schwetzingen (admin_level=8, id=454165, zip=68723)
│   └── Mühlhausen (admin_level=8, id=437377)
├── Heidelberg (admin_level=6, id=285864, zip=69120)
└── Karlsruhe (admin_level=6, id=62518)
Hessen (admin_level=4, id=62650)
├── Bensheim (admin_level=8, id=537055, zip=64625)
└── Odenwaldkreis (admin_level=6, id=62596)
Berlin (admin_level=4, id=62422, zip=10178)


In [84]:
from anytree import Node, RenderTree
import overpy
import random

# Initialize the Overpass API
api = overpy.Overpass()

def get_postal_code(relation_id, api):
    """
    Given a relation_id (e.g., of your newly sampled node),
    run an Overpass query to fetch a random postal code
    boundary within this area. Returns the postal code
    string if found, otherwise None.
    """
    query = f"""
    [out:json];
    relation({relation_id})->.city;
    .city map_to_area -> .cityArea;
    relation[boundary=postal_code](area.cityArea);
    out center;
    """
    try:
        result= api.query(query)
        if result.relations:
            sample = random.choice(result.relations)
            return sample.tags.get('postal_code')
        else:
            return None
    except Exception as e:
        logging.error(f"Error retrieving postal code for relation_id={relation_id}: {e}")
        return None
            
def sample_child_relation(parent_id, child_admin_level, api):
    """
    Find a random child relation within a specified parent relation area.
    Supports both administrative boundaries and streets (admin_level=99).
    In scenarios where no relation is found for the specified administrative level within the area, the function dynamically increments the admin level and continues the search until a valid result is retrieved. 

    Parameters
    ----------
    parent_id : int
        The OSM relation ID of the parent area.
    child_admin_level : int
        The administrative level to search for (99 for streets).
    api : overpy.Overpass
        An instance of the Overpass API.

    Returns
    -------
    tuple
        (child_name, child_id) if found, else (None, None).
    """
    # Define query templates
    queries = {
        'street': """
            [out:json];
            relation({parent_id})->.parent;
            .parent map_to_area -> .parentArea;
            way["highway"]["name"](area.parentArea);
            out center;
        """,
        'admin': """
            [out:json];
            relation({parent_id})->.parent;
            .parent map_to_area -> .parentArea;
            relation["boundary"="administrative"]
                   ["admin_level"="{child_level}"]
                   (area.parentArea);
            out center;
        """
    }

    while True:
        # Select appropriate query based on admin_level
        is_street = child_admin_level == 99
        query = queries['street'] if is_street else queries['admin']
        
        # Format query with parameters
        formatted_query = query.format(parent_id=parent_id,child_level=child_admin_level)
        try:
            # Execute query
            result = api.query(formatted_query)

            # Process results based on type
            if is_street and result.ways:
                sample = random.choice(result.ways)
                return sample.tags.get('name'), sample.id
            elif not is_street and result.relations:
                sample = random.choice(result.relations)
                return sample.tags.get('name'), sample.id
            
            # If no results, increment admin level and try again
            child_admin_level += 1

        except Exception as e:
            logging.error(f"Error in sample_child_relation: {e}")
            return None, None

def rebuild_tree(root_node):
    """
    Rebuild a tree structure starting from the given root_node.

    Parameters
    ----------
    root_node : Node
        The root node of the tree to rebuild.

    Returns
    -------
    list
        A list of new root nodes for the rebuilt tree.
    """
    new_nodes = []  

    def rebuild_subtree(original_node, new_parent=None):
        """
        Recursively rebuild the subtree for the given original_node.
        Attach the new node to new_parent if provided.
        """
        if new_parent is None:
            # For root node, keep original properties
            new_node = Node(
                original_node.name,
                parent=new_parent,
                admin_level=original_node.admin_level,
                id=original_node.id
            )
            
            new_nodes.append(new_node)
        else:
            # For non-root nodes, sample new ones
            sampled_name, sampled_id = sample_child_relation(new_parent.id, original_node.admin_level, api)
            if sampled_name and sampled_id:
                new_node = Node(
                    sampled_name,
                    parent=new_parent,
                    admin_level=original_node.admin_level,
                    id=sampled_id
                )
                # If the original node had a zip attribute, fetch a postal code for the new_node
                if hasattr(original_node, 'zip'):
                    postal_code = get_postal_code(new_node.id, api)
                    if postal_code:
                        new_node.zip = postal_code
            else:
                return  # Skip if no sample found

        # Recurse for all children of the current node
        for child in original_node.children:
            rebuild_subtree(child, new_parent=new_node)

    # Start rebuilding the tree from the given root_node
    rebuild_subtree(root_node)
    return new_nodes

# Rebuild and print the tree for each root node
new_roots = [tree_root for root in old_roots for tree_root in rebuild_tree(root)]
for new_root in new_roots:
    for pre, _, node in RenderTree(new_root):
        # Print formatted tree with admin level and ID for better readability
        zip_info = f", zip={node.zip}" if hasattr(node, 'zip') else ""
        print(f"{pre}{node.name} (admin_level={node.admin_level}, id={node.id}{zip_info})")


Baden-Württemberg (admin_level=4, id=62611)
├── Ulm (admin_level=6, id=62495)
│   ├── Einsingen (admin_level=8, id=16691167)
│   │   └── Seidelbastweg (admin_level=99, id=22744548)
│   ├── Jungingen (admin_level=8, id=16691172)
│   └── Einsingen (admin_level=8, id=16691167)
├── Hohenlohekreis (admin_level=6, id=62394, zip=74679)
└── Landkreis Heidenheim (admin_level=6, id=2812850)
Hessen (admin_level=4, id=62650)
├── Wolfhagen (admin_level=8, id=452969)
└── Lahn-Dill-Kreis (admin_level=6, id=62368)
Berlin (admin_level=4, id=62422)


In [85]:
from anytree import PreOrderIter

def map_trees(old_roots, new_roots):
    """
    Maps nodes between two tree structures using pre-order traversal. Assumes the tree structures are identical in terms of shape and traversal order.
    Parameters
    ----------
    old_roots : list
        List of root nodes from the original trees
    new_roots : list 
        List of root nodes from the new trees
    Returns
    -------
    dict
        Dictionary mapping node names from old trees to corresponding node names in new trees
    """
    if len(old_roots) != len(new_roots):
        raise ValueError("The two lists must have the same number of root nodes.")

    mapping = {}

    for old_root, new_root in zip(old_roots, new_roots):
        iter1 = PreOrderIter(old_root)
        iter2 = PreOrderIter(new_root)

        for node1, node2 in zip(iter1, iter2):
            mapping[node1.name] = node2.name
            if hasattr(node1, 'zip') and hasattr(node2, 'zip'):
                mapping[node1.zip] = node2.zip
    
    return mapping

# Example Usage
mapping = map_trees(old_roots, new_roots)
print(mapping)



{'Baden-Württemberg': 'Baden-Württemberg', 'Rhein-Neckar-Kreis': 'Ulm', 'Wiesloch': 'Einsingen', 'Bergstraße': 'Seidelbastweg', 'Schwetzingen': 'Jungingen', 'Mühlhausen': 'Einsingen', 'Heidelberg': 'Hohenlohekreis', '69120': '74679', 'Karlsruhe': 'Landkreis Heidenheim', 'Hessen': 'Hessen', 'Bensheim': 'Wolfhagen', 'Odenwaldkreis': 'Lahn-Dill-Kreis', 'Berlin': 'Berlin'}


In [53]:
import overpy
import random

# Initialize Overpass API
api = overpy.Overpass()

# Step 1: Retrieve the coordinates of a given relation
relation_id = 453394
query_relation = f"""
[out:json][timeout:25];
relation({relation_id});
out center;
"""
result_relation = api.query(query_relation)
lat = result_relation.relations[0].center_lat
lon = result_relation.relations[0].center_lon

# Step 2: Retrieve postal codes around the retrieved coordinates and pick one randomly
radius = 10000  # in meters
query_postal_codes = f"""
[out:json][timeout:25];
relation[boundary=postal_code](around:{radius},{lat},{lon});
out center;
"""
result_postal_codes = api.query(query_postal_codes)
postal_codes = [rel.tags.get("postal_code") for rel in result_postal_codes.relations if "postal_code" in rel.tags]

random_postal_code = random.choice(postal_codes)
print(random_postal_code)


69226


In [162]:
import overpy
# Initialize the Overpass API
api = overpy.Overpass()

# Replace "Heidelberg" with the desired city name
parent_id = 1231959

# Overpass QL query to get basic information about a city
query = f"""
[out:json];
relation({parent_id})->.parent;

// convert relations object into an area object
.parent map_to_area -> .parentArea;

// Find all relations that are a member of this area
way["highway"]["name"](area.parentArea)->.strasse;

// Output the neighboring relations
way.strasse;
out center;
"""

# Run the query
result = api.query(query)


if result.ways:
    sample_relation = random.choice(result.ways)
    child_name = sample_relation.tags.get('name')
    child_id = sample_relation.id
    print(f"Sample Child Name: {child_name}")

Sample Child Name: Brechgasse


# Initialize Tree

In [37]:
import overpy
from anytree import Node, RenderTree

def initialize_tree(location_countries, api):
    """
    Initialize a hierarchy of administrative nodes for the specified countries.

    Germany is initialized with predefined admin_level 4 states for efficiency. 
    For other countries, administrative subdivisions (admin_level 4) are dynamically 
    queried from OpenStreetMap (OSM).

    Parameters
    ----------
    location_countries : list of str
        Names of countries to include in the hierarchy.
    api : overpy.Overpass
        Overpass API instance for querying OSM data.

    Returns
    -------
    dict
        A dictionary {relation_id: Node}, representing the hierarchy of countries and 
        their subdivisions, with each country as a root node.
    """
    hierarchy_dict = {}

    for country_name in location_countries:
        if country_name == "Deutschland":
            # Create root node for Germany
            germany_id = 51477
            root = Node('Deutschland', admin_level=2, id=germany_id)
            hierarchy_dict[germany_id] = root

            # Add predefined child nodes for German states
            states = [
                ("Mecklenburg-Vorpommern", 28322),
                ("Schleswig-Holstein", 51529),
                ("Rheinland-Pfalz", 62341),
                ("Thüringen", 62366),
                ("Saarland", 62372),
                ("Berlin", 62422),
                ("Sachsen", 62467),
                ("Brandenburg", 62504),
                ("Sachsen-Anhalt", 62607),
                ("Baden-Württemberg", 62611),
                ("Hessen", 62650),
                ("Bremen", 62718),
                ("Nordrhein-Westfalen", 62761),
                ("Niedersachsen", 62771),
                ("Hamburg", 62782),
                ("Bayern", 2145268),
            ]
            for state_name, state_id in states:
                state_node = Node(state_name, parent=root, admin_level=4, id=state_id)
                hierarchy_dict[state_id] = state_node

        else:
            # Query to get the country by name
            query_country = f"""
            [out:json];
            relation["name:de"="{country_name}"]["admin_level"="2"]["boundary"="administrative"];
            out body;
            >;
            out skel qt;
            """

            result_country = api.query(query_country)

            if result_country.relations:
                country_relation = result_country.relations[0]
                country_id = country_relation.id
                # Create root node for the country
                root = Node(country_name, admin_level=2, id=country_id)
                hierarchy_dict[country_id] = root

                # Query to get the states of the country
                query_states = f"""
                [out:json];
                relation({country_id})->.parent;
                .parent map_to_area -> .parentArea;

                relation["boundary"="administrative"]["admin_level"="4"](area.parentArea)->.states;

                relation.states;
                out center;
                """

                result_states = api.query(query_states)

                # Add each state as a child node
                for rel in result_states.relations:
                    state_id = rel.id
                    state_name = rel.tags.get("name:de", rel.tags.get("name", "Unknown"))
                    state_node = Node(state_name, parent=root, admin_level=4, id=state_id)
                    hierarchy_dict[state_id] = state_node

    return hierarchy_dict

# Example usage
location_countries = ["Deutschland", "Österreich", "Italien"]
api = overpy.Overpass()

# Initialize the tree and get a dictionary of nodes
hierarchy_dict = initialize_tree(location_countries, api)

# Extract root nodes (those without parents)
root_nodes = [node for node in hierarchy_dict.values() if node.is_root]

# Render the tree for each root node
for root in root_nodes:
    print(RenderTree(root))

Node('/Deutschland', admin_level=2, id=51477)
├── Node('/Deutschland/Mecklenburg-Vorpommern', admin_level=4, id=28322)
├── Node('/Deutschland/Schleswig-Holstein', admin_level=4, id=51529)
├── Node('/Deutschland/Rheinland-Pfalz', admin_level=4, id=62341)
├── Node('/Deutschland/Thüringen', admin_level=4, id=62366)
├── Node('/Deutschland/Saarland', admin_level=4, id=62372)
├── Node('/Deutschland/Berlin', admin_level=4, id=62422)
├── Node('/Deutschland/Sachsen', admin_level=4, id=62467)
├── Node('/Deutschland/Brandenburg', admin_level=4, id=62504)
├── Node('/Deutschland/Sachsen-Anhalt', admin_level=4, id=62607)
├── Node('/Deutschland/Baden-Württemberg', admin_level=4, id=62611)
├── Node('/Deutschland/Hessen', admin_level=4, id=62650)
├── Node('/Deutschland/Bremen', admin_level=4, id=62718)
├── Node('/Deutschland/Nordrhein-Westfalen', admin_level=4, id=62761)
├── Node('/Deutschland/Niedersachsen', admin_level=4, id=62771)
├── Node('/Deutschland/Hamburg', admin_level=4, id=62782)
└── Node('/

In [ ]:
# Filter nodes with admin level 6 or higher
need admin level for street node aswell so i can traverse the tree and generate an appropriate surrogate based on the tree structure. maybe make a new tree with only admin level 6 or higher and street as admin level 99 and then i only need the root node id because then i will search for a fitting surrogate for that root node and then all of the following surrogates are only based on being contained by the parent node and their admin level.
filtered_nodes = {node for node in hierarchy_nodes.items()}
print(hierarchy_nodes)






Admin Level: 2
  Locations: Deutschland, Österreich, Italien

Admin Level: 4
  Locations: Berlin, Baden-Württemberg

Admin Level: 6
  Locations: Heidelberg, Karlsruhe, Rhein-Neckar-Kreis, Odenwaldkreis

Admin Level: 8
  Locations: Wiesloch, Schwetzingen, Bensheim

Starting to build hierarchy...


Processing locations at admin level 8...
Wiesloch 453399 8
Processing 'Wiesloch' at admin level 8 (ID: 453399)...
  Checking for parents at admin level 6...
    Checking if 'Wiesloch' is a child of 'Heidelberg' (ID: 285864)...
    Checking if 'Wiesloch' is a child of 'Karlsruhe' (ID: 62518)...
    Checking if 'Wiesloch' is a child of 'Rhein-Neckar-Kreis' (ID: 62577)...
Processing 'Rhein-Neckar-Kreis' at admin level 6 (ID: 62577)...
  Checking for parents at admin level 4...
    Checking if 'Rhein-Neckar-Kreis' is a child of 'Berlin' (ID: 62422)...
    Checking if 'Rhein-Neckar-Kreis' is a child of 'Baden-Württemberg' (ID: 62611)...
Processing 'Baden-Württemberg' at admin level 4 (ID: 62611)...
  Checking for parents at admin level 2...
    Checking if 'Baden-Württemberg' is a child of 'Deutschland' (ID: 51477)...
Processing 'Deutschland' at admin level 2 (ID: 51477)...
  'Deutschland' is at admin_level 2, treating it as a root node. No parent search needed.
    Added 'Baden-Württemberg' as child of 'Deutschland'
    Added 'Rhein-Neckar-Kreis' as child of 'Baden-Württemberg'
    Added 'Wiesloch' as child of 'Rhein-Neckar-Kreis'
Schwetzingen 454165 8
Processing 'Schwetzingen' at admin level 8 (ID: 454165)...
  Checking for parents at admin level 6...
    Checking if 'Schwetzingen' is a child of 'Heidelberg' (ID: 285864)...
    Checking if 'Schwetzingen' is a child of 'Karlsruhe' (ID: 62518)...
    Checking if 'Schwetzingen' is a child of 'Rhein-Neckar-Kreis' (ID: 62577)...
    Added 'Schwetzingen' as child of 'Rhein-Neckar-Kreis'
Bensheim 537055 8
Processing 'Bensheim' at admin level 8 (ID: 537055)...
  Checking for parents at admin level 6...
    Checking if 'Bensheim' is a child of 'Heidelberg' (ID: 285864)...
    Checking if 'Bensheim' is a child of 'Karlsruhe' (ID: 62518)...
    Checking if 'Bensheim' is a child of 'Rhein-Neckar-Kreis' (ID: 62577)...
    Checking if 'Bensheim' is a child of 'Odenwaldkreis' (ID: 62596)...
  Checking for parents at admin level 4...
    Checking if 'Bensheim' is a child of 'Berlin' (ID: 62422)...
    Checking if 'Bensheim' is a child of 'Baden-Württemberg' (ID: 62611)...
  Checking for parents at admin level 2...
    Checking if 'Bensheim' is a child of 'Deutschland' (ID: 51477)...
    Added 'Bensheim' as child of 'Deutschland'

Processing locations at admin level 6...
Heidelberg 285864 6
Processing 'Heidelberg' at admin level 6 (ID: 285864)...
  Checking for parents at admin level 4...
    Checking if 'Heidelberg' is a child of 'Berlin' (ID: 62422)...
    Checking if 'Heidelberg' is a child of 'Baden-Württemberg' (ID: 62611)...
    Added 'Heidelberg' as child of 'Baden-Württemberg'
Karlsruhe 62518 6
Processing 'Karlsruhe' at admin level 6 (ID: 62518)...
  Checking for parents at admin level 4...
    Checking if 'Karlsruhe' is a child of 'Berlin' (ID: 62422)...
    Checking if 'Karlsruhe' is a child of 'Baden-Württemberg' (ID: 62611)...
    Added 'Karlsruhe' as child of 'Baden-Württemberg'
Rhein-Neckar-Kreis 62577 6
Odenwaldkreis 62596 6
Processing 'Odenwaldkreis' at admin level 6 (ID: 62596)...
  Checking for parents at admin level 4...
    Checking if 'Odenwaldkreis' is a child of 'Berlin' (ID: 62422)...
    Checking if 'Odenwaldkreis' is a child of 'Baden-Württemberg' (ID: 62611)...
  Checking for parents at admin level 2...
    Checking if 'Odenwaldkreis' is a child of 'Deutschland' (ID: 51477)...
    Added 'Odenwaldkreis' as child of 'Deutschland'

Processing locations at admin level 4...
Berlin 62422 4
Processing 'Berlin' at admin level 4 (ID: 62422)...
  Checking for parents at admin level 2...
    Checking if 'Berlin' is a child of 'Deutschland' (ID: 51477)...
    Added 'Berlin' as child of 'Deutschland'
Baden-Württemberg 62611 4

Processing locations at admin level 2...
Deutschland 51477 2
Österreich 16239 2
Processing 'Österreich' at admin level 2 (ID: 16239)...
  'Österreich' is at admin_level 2, treating it as a root node. No parent search needed.
Italien 365331 2
Processing 'Italien' at admin level 2 (ID: 365331)...
  'Italien' is at admin_level 2, treating it as a root node. No parent search needed.

Hierarchy building complete.
Deutschland (ID: 51477)
├── Baden-Württemberg (ID: 62611)
│   ├── Rhein-Neckar-Kreis (ID: 62577)
│   │   ├── Wiesloch (ID: 453399)
│   │   └── Schwetzingen (ID: 454165)
│   ├── Heidelberg (ID: 285864)
│   └── Karlsruhe (ID: 62518)
├── Bensheim (ID: 537055)
├── Odenwaldkreis (ID: 62596)
└── Berlin (ID: 62422)
Österreich (ID: 16239)
Italien (ID: 365331)

In [102]:
import overpy
# Initialize the Overpass API
api = overpy.Overpass()

parent_id = 62422

# Overpass QL query to get basic information about a city
query = f"""
[out:json];
relation({parent_id})->.parent;
.parent map_to_area -> .parentArea;

// Find all relations that are a member of this area
way["highway"]["name"](area.parentArea)->.strasse;

// Output the neighboring relations
way.strasse;
out center;
"""

# Run the query
result = api.query(query)

# Process relations
for ways in result.ways:
    # print(f"Relation ID: {ways.id}")
    # print(f"Latitude: {ways.center_lat}")
    # print(f"Longitude: {ways.center_lon}")
    # print("Tags:")
    # for tag_key, tag_value in ways.tags.items():
    #     print(f"  {tag_key}: {tag_value}")
    # print("\n")
    print(f"  Name: {ways.tags['name']}")

  Name: Ursula-Goetze-Straße
  Name: Hönower Straße
  Name: Gundelfinger Straße
  Name: Frankfurter Allee
  Name: Am Wiesengrund
  Name: Hauptstraße
  Name: Michael-Brückner-Straße
  Name: Olympische Straße
  Name: Jafféstraße
  Name: Müllerstraße
  Name: Köpenicker Landstraße
  Name: Alt-Reinickendorf
  Name: Liebknechtbrücke
  Name: Barkenhof
  Name: Kienhorststraße
  Name: Alte Jafféstraße
  Name: Rhinstraße
  Name: Lindauer Allee
  Name: Wiesenschlag
  Name: Alt-Reinickendorf
  Name: Hosemannstraße
  Name: Kastanienallee
  Name: Bürknerstraße
  Name: Halemweg
  Name: Messedamm
  Name: Heerstraße
  Name: Straße des 17. Juni
  Name: Elsenbrücke
  Name: Scharnweberstraße
  Name: Mehringdamm
  Name: Bulgarische Straße
  Name: Roedernallee
  Name: Warschauer Straße
  Name: Heilmannring
  Name: Reichsstraße
  Name: Alt-Friedrichsfelde
  Name: Wisbyer Straße
  Name: Masurenallee
  Name: Kesperhof
  Name: Humboldtstraße
  Name: Heckerdamm
  Name: Schlehdornsteig
  Name: Alt-Friedrichsfelde